### PR TITLE
refactor(agent): improve magic file recognition performance

### PR DESCRIPTION
**Pull request by lavarou**
Speed up speed library/framework detection by performing a suffix match on the 'magic' file pattern with case insensitive string comparison instead of a substring search within a lowercased filename. This is possible because all of the 'magic' file search patterns patterns are right anchored.

### CODE REVIEW

File: agent/php_execute.c [hunk 1]
```diff
@@ -705,7 +713,8 @@ static void nr_php_show_exec_return(NR_EXECUTE_PROTO TSRMLS_DC) {
 static nrframework_t nr_try_detect_framework(
     const nr_framework_table_t frameworks[],
     size_t num_frameworks,
-    const char* filename TSRMLS_DC);
+    const char* filename,
+    const size_t filename_len TSRMLS_DC);
```
Comment (bduranleau-nr) [inline agent/php_execute.c:717]:
```suggestion
    const size_t filename_len);
```

File: agent/php_execute.c [hunk 2]
```diff
@@ -765,16 +774,17 @@ void nr_framework_create_metric(TSRMLS_D) {
...
     return;
   }
 
   if (NR_FW_UNSET == NRINI(force_framework)) {
     nrframework_t detected_framework = NR_FW_UNSET;
 
-    detected_framework = nr_try_detect_framework(frameworks, num_frameworks,
-                                                 filename TSRMLS_CC);
+    detected_framework = nr_try_detect_framework(
+        frameworks, num_frameworks, filename, filename_len TSRMLS_CC);
```
Comment (bduranleau-nr) [inline agent/php_execute.c:787]:
```suggestion
        frameworks, num_frameworks, filename, filename_len);
```

File: agent/php_execute.c [hunk 3]
```diff
@@ -800,13 +812,14 @@ static void nr_execute_handle_framework(const nr_framework_table_t frameworks[],
 static nrframework_t nr_try_detect_framework(
     const nr_framework_table_t frameworks[],
     size_t num_frameworks,
-    const char* filename TSRMLS_DC) {
+    const char* filename,
+    const size_t filename_len TSRMLS_DC) {
```
Comment (bduranleau-nr) [inline agent/php_execute.c:816]:
```suggestion
    const size_t filename_len) {
```

File: agent/php_execute.c [hunk 4]
```diff
@@ -869,12 +881,13 @@ static nrframework_t nr_try_force_framework(
   return NR_FW_UNSET;
 }
 
-static void nr_execute_handle_library(const char* filename TSRMLS_DC) {
-  char* filename_lower = nr_string_to_lowercase(filename);
+static void nr_execute_handle_library(const char* filename,
+                                      const size_t filename_len TSRMLS_DC) {
```
Comment (bduranleau-nr) [inline agent/php_execute.c:885]:
```suggestion
                                      const size_t filename_len) {
```

File: agent/php_execute.c [hunk 5]
```diff
@@ -886,18 +899,17 @@ static void nr_execute_handle_library(const char* filename TSRMLS_DC) {
...
-
-  nr_free(filename_lower);
 }
 
-static void nr_execute_handle_logging_framework(
-    const char* filename TSRMLS_DC) {
-  char* filename_lower = nr_string_to_lowercase(filename);
+static void nr_execute_handle_logging_framework(const char* filename,
+                                                const size_t filename_len
+                                                    TSRMLS_DC) {
```
Comment (bduranleau-nr) [inline agent/php_execute.c:906]:
```suggestion
                                                const size_t filename_len) {
```

File: agent/php_execute.c [hunk 6]
```diff
@@ -924,12 +936,20 @@ static void nr_execute_handle_logging_framework(
...
  */
-static void nr_php_user_instrumentation_from_file(
-    const char* filename TSRMLS_DC) {
-  nr_execute_handle_framework(all_frameworks, num_all_frameworks,
-                              filename TSRMLS_CC);
-  nr_execute_handle_library(filename TSRMLS_CC);
-  nr_execute_handle_logging_framework(filename TSRMLS_CC);
+static void nr_php_user_instrumentation_from_file(const char* filename,
+                                                  const size_t filename_len
+                                                      TSRMLS_DC) {
```
Comment (bduranleau-nr) [inline agent/php_execute.c:941]:
```suggestion
                                                  const size_t filename_len) {
```

File: agent/php_execute.c [hunk 7]
```diff
@@ -948,7 +969,7 @@ static void nr_php_execute_file(const zend_op_array* op_array,
   /*
    * Check for, and handle, frameworks and libraries.
    */
-  nr_php_user_instrumentation_from_file(filename TSRMLS_CC);
+  nr_php_user_instrumentation_from_file(filename, filename_len TSRMLS_CC);
```
Comment (bduranleau-nr) [inline agent/php_execute.c:972]:
```suggestion
  nr_php_user_instrumentation_from_file(filename, filename_len);
```

File: agent/php_execute.c
Comment (bduranleau-nr) [inline agent/php_execute.c:1]: Not necessarily something that needs to be fixed, but why the extra space between every `NR_PSTR` and the `("file_arg")`?

File: axiom/util_strings.h [hunk 8]
```diff
@@ -438,4 +439,36 @@ static inline int nr_toupper(int c) {
...
+  if (slen < pattern_len) {
+    /* input shorter than pattern */
+    return false;
+  }
+
+  if (NULL == s) {
+    /* invalid input */
+    return false;
+  }
+
```
Comment (bduranleau-nr) [inline axiom/util_strings.h:468]: What happens if `pattern` is `NULL`?

Comment (lavarou) [reply axiom/util_strings.h:468]: `nr_stricmp` handles this case and returns `1` which will result in `nr_striendswith` returning `false`. I will add a test case for this.

File: agent/php_execute.c
Comment (lavarou) [reply agent/php_execute.c:1]: No reason other than that's what search and replace did. I can remove the extra space if it's bothering.

**Comment by lavarou** [comment]
@bduranleau-nr I really don't think removal of `TSRM_*` macros belongs to this PR.

File: axiom/util_strings.h (hunk 8 above)
Comment (lavarou) [reply axiom/util_strings.h:468]: See 962c98c68f28cef52cd8394010b59c16d8798e02

File: agent/php_execute.c
Comment (lavarou) [reply agent/php_execute.c:1]: Does it look better after 0e728b3a8ac7d838fac0e00c9094d25ff9d68ecc?

Comment (bduranleau-nr) [reply agent/php_execute.c:1]: Great! Thank you!

**Comment by bduranleau-nr** [comment]
> @bduranleau-nr I really don't think removal of `TSRM_*` macros belongs to this PR - let's try to keep things focused.

Understood, that isn't explicitly part of this effort but I do think it's a good idea to kill two birds with one stone and remove references to `TSRM_*` if the function call/function signature is already being modified. 
Either way it isn't a blocker for me so I'll go ahead and approve the PR and leave decision that up to you.

**Review by bduranleau-nr — APPROVED** [review]

File: agent/php_execute.c (hunk 1 above)
Comment (lavarou) [reply agent/php_execute.c:717]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c [hunk 9]
```diff
@@ -765,16 +774,17 @@ void nr_framework_create_metric(TSRMLS_D) {
  */
 static void nr_execute_handle_framework(const nr_framework_table_t frameworks[],
                                         size_t num_frameworks,
-                                        const char* filename TSRMLS_DC) {
+                                        const char* filename,
+                                        const size_t filename_len TSRMLS_DC) {
```
[the remark this answers is not in the document]
Comment (lavarou) [reply agent/php_execute.c:778]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 2 above)
Comment (lavarou) [reply agent/php_execute.c:787]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 3 above)
Comment (lavarou) [reply agent/php_execute.c:816]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 4 above)
Comment (lavarou) [reply agent/php_execute.c:885]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 5 above)
Comment (lavarou) [reply agent/php_execute.c:906]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 6 above)
Comment (lavarou) [reply agent/php_execute.c:941]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c (hunk 7 above)
Comment (lavarou) [reply agent/php_execute.c:972]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

File: agent/php_execute.c [hunk 10]
```diff
@@ -1717,8 +1739,9 @@ void nr_php_user_instrumentation_from_opcache(TSRMLS_D) {
     (void)val;
 
     filename = ZEND_STRING_VALUE(key_str);
+    filename_len = ZEND_STRING_LEN(key_str);
 
-    nr_php_user_instrumentation_from_file(filename TSRMLS_CC);
+    nr_php_user_instrumentation_from_file(filename, filename_len TSRMLS_CC);
```
[the remark this answers is not in the document]
Comment (lavarou) [reply agent/php_execute.c:1744]: See https://github.com/newrelic/newrelic-php-agent/pull/766#issuecomment-1816887018

**Review by zsistla — APPROVED** [review]

**Pull request merged**